### ISSUE TITLE
Nightly build fails for queryprocessing_test

### ISSUE THREAD

**Issue by alemoi** [Starcounter/Starcounter#3429]
Here is the place in the latest nightly:

https://teamcity.starcounter.org/viewLog.html?tab=buildLog&logTab=tree&filter=debug&expand=all&buildId=1536#_focus=6387

**Comment by alexanderstc** [comment]
I'll try to reproduce.

**Comment by alexanderstc** [comment]
Oh, it's develop, can't help much there.

**Comment by k-rus** [comment]
The builds started to fail after changes from @alemoi, @warpech and @per-samuelsson 
It seems to be related to the changed number of tables affecting the test. What exactly was changed I don't know. So it is question to contributors.

**Comment by k-rus** [comment]
By the way, does failed build suppose to be available for download? All nightly builds fail on the build server, but they are available for download on the download web page. Is it correct?

**Comment by per-samuelsson** [comment]
I took a look at it locally to first of all be able to see *what* fails, and it's again one of those tests that expect a fixed number of entries for a result:

This line fails in metadata tests:

```
Trace.Assert(count == 49);
```

The count comes from counting the result of `select rv from rawView rv`, and when I run this test locally, it's 52, not 49.

Not easiest thing to understand what functionality is really tested by this, and I don't know how to check that "49", and how I can verify it to find those 3 extras that are now part of that set.

How should I continue here @k-rus? Where do the 49 come from? Is it a count you've done locally when you wrote that test?

**Comment by per-samuelsson** [comment]
OK, I made a custom build to just get somewhere with this, and here are the 52 tables that this test consider a failure - because the test statically has specified that there could only be 49 tables (but not specified which):

![image](https://cloud.githubusercontent.com/assets/2902045/13567653/c4e6ba08-e45c-11e5-8797-3f10c7a85ed0.png)

The test say it should be 49. There are 52. Who's to blame for that, I don't know. I guess no one can tell except @k-rus, and he says it's probably me to blame, so we are in some situation here where I don't know how to get further and how to fix this.

Yes, I can just change 49 to 52. But I have no idea what that indicates, so I won't.

Just generally, I expressed quite some time ago that I think [these kind of tests serve very little purpose](https://github.com/Starcounter/Starcounter/issues/3042#issuecomment-156072013), and I'm a bit frustrated I have to spend time trying to figure out why these numbers doesn't match, when the relevance of them is (IMO) close to zero. Again, IMO, these tests don't help us - quite the opposite actually, since they are rather hard to diagnose, and just steal time that I'd rather spend on something else.

Would be great if you could help me fix this eventually, @k-rus, because I'm quite confused what is actually wrong here. Thanks.

**Comment by alemoi** [comment]
Per I will check it. Might be related to my changes in Starcounter.Extensions library, however I was sure its excluded like EditionLibraries from the test. Thank you for initial investigation though!

**Comment by k-rus** [comment]
> The count comes from counting the result of select rv from rawView rv, and when I run this test locally, it's 52, not 49.
> 
> Not easiest thing to understand what functionality is really tested by this, and I don't know how to check that "49", and how I can verify it to find those 3 extras that are now part of that set.
> 
> How should I continue here @k-rus? Where do the 49 come from? Is it a count you've done locally when you wrote that test?

The tests assume that the number of tables in the test database is constant. Since the test was not changed, then number of meta-tables or some other unexpected by test tables were changed. The list of tables shows unexpected tables in namespace `Extensions`.

I didn't know who's changes were responsible those I mentioned everyone related to the build. Now, @alemoi will check the problem.

Note, this test is not so any more in pnext.

**Comment by per-samuelsson** [comment]
> I didn't know who's changes were responsible those I mentioned everyone related to the build.

From #3423, directed to me:

> note develop nightly failed **probably due to your changes**

And about this:

> Now, @alemoi will check the problem.

I know he is. And before that, I spent time on it. You are not at all listening to my essential message here: if there is anything to be considered a bug in this issue, it's how the test is designed. But lets not have this discussion on a tracker. If you ever feel interested, I'll can try explain again to you why I think it is. Just let me know.

**Comment by alemoi** [comment]
We have discussed the issue with the test yesterday. Apparently the problem is more general: when adding some classes, the kernel can change the order in which results are returned and there are cases when there is no sorting on that column, so some tests may suffer from this re-ordering issue. My thinking: I don't think its possible to avoid such problems but we should not run the tests in limited environment (like disabling the `EditionLibraries` folder), I think we should solve this ordering problems inside the tests unfortunately once they appear. As @k-rus pointed the issue with number of classes is not in pnext queryprocessing anymore, which is good. I also think if pnext has no such limitations we should consider running the queryprocessing tests there without "DisableEditionLibraries" flag and maybe even remove this flag completely, same as "NoNetworkGateway" and some other flags that test "incomplete" Starcounter.

**Comment by per-samuelsson** [comment]
>  [...] without "DisableEditionLibraries" flag and maybe even remove this flag completely, same as "NoNetworkGateway" and some other flags that test "incomplete" Starcounter.

And `NoDb`, which I don't think is used anymore either? I totally agree with your reasoning - these flags that is not actively used, we should remove, and all tests should operate on "standard Starcounter". After all, that's what our users will use, and hence that's what we want to test (except in very special corner-cases possibly, but then tests should individually adapt to Starcounter, not the other way around).

**Comment by per-samuelsson** [comment]
> We have discussed the issue with the test yesterday. Apparently the problem is more general: when adding some classes, the kernel can change the order in which results are returned and there are cases when there is no sorting on that column, so some tests may suffer from this re-ordering issue.

This test didn't suffer from that. This test has nothing to do with ordering. It assumes there are 49 tables in the database. And if it's not, that's an error. Every time anyone add a new database class in Starcounter, this test will fail.

**Comment by alemoi** [comment]
> This test didn't suffer from that. This test has nothing to do with ordering. It assumes there are 49 tables in the database. And if it's not, that's an error. Every time anyone add a new database class in Starcounter, this test will fail.

True, but as @k-rus points, its not like that in pnext anymore. So, knowing that, this test is like an exception, that we simply should know about and don't spend any more time on. For the future, I really want to force removal of all these flags + not support manual start of sccode, etc (which some of my tests still use, as I remember), only the ubiquitous star.exe! :)

**Comment by k-rus** [comment]
> From #3423, directed to me:
> 
> note develop nightly failed probably due to your changes

Yes, you were one of the three developers, who committed the changes, which broke the nightly build. I didn't know if your changes were breaking or not. I mentioned you because it was related to side discussion there, and you as other two developers didn't react to my notice in this issue. It is great that you took responsibility and look into the issue. Unfortunately, nightly builds are grey zone and it is usually unclear who should investigate failures. This is why I am cautious on reducing number of tests in pnext daily build.

> After all, that's what our users will use, and hence that's what we want to test (except in very special corner-cases possibly, but then tests should individually adapt to Starcounter, not the other way around).

I will repeat what I said and wrote number of times already. Not all tests are system and product tests. There are unit tests and component tests, which tests only certain components without being disturb other components. IMHO, the test environment including regression test must allow running unit and component tests.



> We have discussed the issue with the test yesterday. Apparently the problem is more general: when adding some classes, the kernel can change the order in which results are returned and there are cases when there is no sorting on that column, so some tests may suffer from this re-ordering issue.
> This test didn't suffer from that. This test has nothing to do with ordering. It assumes there are 49 tables in the database. And if it's not, that's an error. Every time anyone add a new database class in Starcounter, this test will fail.

We discussed different test (SQL regression test), because the problem, which we meet with this test (number of tables in the database), is also  met in SQL regression test. Adding new table in the test database might fail SQL regression test.

Another important point adding new meta-table or table, which needed for maintaining Starcounter work, is breaking change, since upgrading Starcounter on existing database without upgrading meta-data might break consistency of the database.

**Comment by k-rus** [comment]
> For the future, I really want to force removal of all these flags + not support manual start of sccode, etc (which some of my tests still use, as I remember), only the ubiquitous star.exe! :)

I don't support this idea. It will limit ability to do component testing and debugging.
